---
title:  "<big><u>IMAP-PART 03</u></big><br><br>Bioinformatics Analysis of Microbiome Data<br><br>"
subtitle: |
  | ![](images/bioinformatics.png)<br>
  | IMAP: Integrated Microbiome Analysis Pipelines<br><br><small>[Latest GitHub-Repo](https://github.com/tmbuza/imap-bioinformatics-analysis/)</small>
date:
- Updated on `r Sys.Date()`
author: Teresia Mrema-Buza
output: 
  html_document: 
css:
- styles.css
csl: "library/apa.csl"
bibliography:
- library/references.bib
- library/imap.bib
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-bioinformatics-analysis"
description: |
  Minimal summary describing features for getting started with microbiome data analysis.
---

<div id="main-container">

<!-- Google Fonts: https://fonts.google.com/?preview.text=IMAP%20template&preview.text_type=custom -->
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Anton" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@100;300;400;500;700,900&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Oswald:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Merriweather:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@100;200;300;400;700&display=swap" rel="stylesheet">

<span class="fas fa-my-custom-icon"></span>


```{r echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  cache = FALSE,
  comment = NA,
  fig.path='./figures/',
  fig.show='asis',
  dev = 'png',
  fig.align='center',
  out.width = "70%",
  fig.width = 7,
  fig.asp = 0.7,
  fig.show = "asis"
)

library(tidyverse, suppressPackageStartupMessages())
library(schtools)
```


# General Overview

- We use the snakemake workflow management system [@Koster2021; @Snakemake2023] for:
  - Maintaining reproducibility in technical validation and regeneration of results.
  - Creating scalable data analysis scaled to server, grid or cloud environment.
  - Fostering sustainable improvement of the microbiome data analysis.
- We also review existing workflows [@Snakemake2023; @Mothursnakemake] to help in gaining a better insights for improving microbiome data analysis.
- We break any complex workflows into small contiguous but related chunks where each major step form a separate executable snakemake rule.

<br>

> We envision to keep fostering on continuous integration and development of highly reproducible workflows.

<img src="images/cicd.png">


<br><br>

## Current project tree
```{bash tree, include=FALSE}
bash workflow/scripts/tree.sh
```

```{bash treeimg}
cat results/project_tree.txt
```

<br><br>

##  Current snakemake workflow
- Typically the snakemake workflow is defined by specifying rules.
- The rule-graph graphically shows the input-output files. 
- The snakemake is capable of automatically determining the dependencies between the rules and creates a dot-like DAG (Directed Acyclic Graph).

```{bash ruledag, include=FALSE}
bash workflow/scripts/rules_dag.sh
```

<img src="dags/rulegraph.svg">

<br><hr width=100%><br>


## Screenshot of interactive snakemake report {#smkreport}
> The snakemake html report can be viewed using any compartible browser, such as chrome to explore more on the workflow and the associated statistics. You will be able to close the left bar to get a better wider view of the display.

<img src="images/smkreport/screenshot.png">

<br><hr width=100%><br>


# Microbial Profiling Using Mothur

## Mothur reference databases
<ol>
  <li>Mothur-based SILVA reference files [@MothurSILVA].</li>
  <li>Mothur-based RDP reference files [@MothurRDP]. Note: The RDP database is to classify 16S rRNA gene sequences to the genus level.</li>
  <li>ZymoBIOMICS Microbial Community Standard (Cat # D6306)[@Zymo2023]. The ZymoBIOMICS Microbial Community DNA Standard is designed to assess bias, errors and other artifacts after the step of nucleic acid purification.</li>
</ol>

There are four methods that can be used to profile microbial communities present in a sample. Here we briefly decribe each method:

## 1.Classify OTUs
- OTUs (Operational Taxonomic Units (OTUs)) are clusters of similar sequences and are commonly accepted as analytical units in microbial profiling when using 16S rRNA gene markers.

## 2. Classify Phylotypes
- A phylotype in microbiome research is a DNA sequence or group of sequences sharing more than an arbitrarily chosen level of similarity of a 16S rRNA gene marker.

## 3. Classify ASVs
- ASVs Amplicon Sequence Variants (ASVs)in microbiome research is any inferred single DNA sequences recovered from a bioinformatics analysis of 16S rRNA marker genes.
- ASV is typically really a cluster of sequences that are one or two bases apart from each other.

## 4. Classify Phylogenies
- Microbial phylogenies are from gene sequence homologies.  Models of mutation determine the most-likely evolutionary histories.

<br><br>

# Preliminary analysis using `Mothur`
The preliminary analysis (`alpha_beta_diversity rule`) is part of the bioinformatics analysis. It includes:

- Creating reads count for each group.
- Subsampling for downstream analysis.
- Rarefaction.
- Computing Alpha diversity metrics.
- Computing Beta diversity metrics.
- Getting sample distances.
- Constructing sample phylip tree.
- Generating ordination matrices including PCoA and NMDS.


<br>

## Citation
  > Please consider citing the [iMAP article](https://rdcu.be/b5iVj) [@TMBuza2019] if you find any part of the IMAP practical user guides helpful in your microbiome data analysis.

<br><hr width=100%><br>

# Appendix {-}

## Troubleshooting (in progress)
<ol>
  <li>Are chimeras removed by default in newer versions on mothur?</li>
    <ul>Yes. Chimeras are removed by default. You can still run the remove.seqs command without error, but it is not necessary. Remove chimera sequence explained [here](https://forum.mothur.org/t/chimera-vsearch-removes-groups-and-generates-empty-files/21342/4)</ul>.
  <li>Mothur dist.seqs taking too long.</li>
  <ul>
    <li>Merged reads are too long, probably over 300pb.</li>
    <li>Reads not overlaping when merging the paired reads.</li>
    <li>Too many uniques representative sequences probably caused by lack of overlapping.</li>
    <li>No enough computer power which suggest a use of HPC or Cluster.</li>
  </ul>
</ol>

<br><br><hr width=50%><br>

## References
::: {#refs}
:::

<br><br>

</div> <!-- end main-container -->


<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`. <br><br>
Github_Repo: `r rmarkdown::metadata$github_repo`.  
Repo compiled and maintained by: [`r rmarkdown::metadata$author`](mailto:`r rmarkdown::metadata$email_address`).  


</div><br><br>
