---
title: "Microbiome Data Analysis Snakemake Workflows<br><br>"
subtitle: |
  | Integrated Solutions for Better Results<br><br>
author: Teresia Mrema-Buza
date: |
  | ![](images/new_horizon.png)<br><br>
  | Updated on `r Sys.time()` <br><br>
csl: "library/apa-numeric-superscript.csl"
bibliography: "library/references.bib"
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-mothur"
github_page: "https://tmbuza.github.io/imap-mothur/"
related_website: "https://complexdatainsights.com"
description: |
  This guide provides integrated and highly curated solutions for achieving better results.
---

<br>

 >Note: We use the snakemake workflow engine [@Koster2021] to maintain reproducibility in technical validation, regeneration of results, and improvement of the microbiome bioinformatics analysis.

<hr width=100%>

<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Montserrat&display=swap" rel="stylesheet">


```{r echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  cache = FALSE,
  comment = NA,
  fig.path='./figures/',
  fig.show='asis',
  dev = 'png',
  fig.align='center',
  out.width = "70%",
  fig.width = 7,
  fig.asp = 0.7,
  fig.show = "asis"
)

library(tidyverse, suppressPackageStartupMessages())
library(mikropml)
library(schtools)

```



```{css echo = FALSE}
/* ############
# CONTAINER CSS
############ */

.author, .subtitle{
  display: none;
} 

.date{
  color: yellow;
} 
.main-container {
  max-width: 80%;
  background-color: black;
  color: white;
  padding: 3% 3% 3% 3%;
  margin-left: auto;
  margin-right: auto;
}
body{
  font-size: 14px;
  text-align: left;
}

#header {
  background-color: none;
  color: white;
  /* line-height: 4; */
  font-family: "Montserrat", sans-serif;
  text-align: center;
  width: 100%;
  padding: 5%;
  text-transform: none;
  font-weight: bold;
  background-image: url("images/bkgd.png");
  background-repeat: no-repeat;
  height: 100%; /* You must set a specified height */
  background-position: center; /* Center the image */
  background-size: cover; /* Resize the background image to cover the entire container */
  
}
```

```{css echo = FALSE} 
/* ############
# BODY CSS
############ */
#header h1{
  font-size: 4em;
  text-transform: capitalize;
  color: black;
}
#header h2{
  font-size: 2.8em;
}

#header h3{
  font-size: 1.8em;
  color: black;
  text-transform: capitalize;
} 
#header img{
width:96%;
}

body {
  background-color: F5F5F5;
  color: white;
  font-family: "Montserrat", sans-serif;
}
body h1{
  font-size: 2.8em;
  text-transform: capitalize;
}
body h2{
  font-size: 2.0em;
}

body h3{
  font-size: 1.6em;
}

table tbody tr td {
  font-size: 14px;
  text-transform: capitalize;
}
body img{
  width: 100%;
  display: block;
  margin-left: auto;
  margin-right: auto;
  padding: 20px, 50px, 20px, 50px;
}
#glimpse{
  text-align: center;
}

#glimpse h1{
  text-transform: none;
}

a{
  color: #6bb7f5;
}
#tablefield {
  max-width: 80%;
  margin-left: auto;
  margin-right: auto;
}
hr {
  display: inline-block;
  text-align: left;
  margin: 0px;
}
#glimpse hr {
  display: inline-block;
  margin: 0px;
  text-align: left;
}

#appendix img{
  width: 75%;
}
```


```{css echo = FALSE} 
/* ############
# MEDIA: max-width: 768px
############ */
@media only screen and (max-width: 768px) {
  .main-container {
  max-width: 100%;
  padding: 1% 1% 1% 1%;
  margin-left: auto;
  margin-right: auto;
  }
  #header h1{
    font-size: 3em;
    text-transform: capitalize;
    color: black;
  }

   body {
    font-size: 12px;
    width: 100%; 
  } 
  #footer {
    text-align: left;
    font-size: 1.1em;
  } 

  #tablefield {
  width: 100%;
  margin-left: auto;
  margin-right: auto;
} 
body h1{
  font-size: 2.5em;
  text-transform: capitalize;
}
body h2{
  font-size: 1.8em;
}

body h3{
  font-size: 1.6em;
}

#tablefield h2{
  font-size: 1.6em;
}

}
```
<br><br>

# Improved Mothur Snakemake Workflows
 
## imap-mothur
- A workflow for microbial profiling using 16S rRNA gene markers.
- We review existing workflows [@mothursnakemake] and create a more reproducible pipeline.
- Each major step form a separate snakemake rule.


<br><br>

## Workflow framework
<img src="dags/rulegraph.svg"/>


```{r mothur_snakemake_dag, eval=FALSE, include=FALSE}
library(DiagrammeR)
library(DiagrammeRsvg)

mothur_snakemake_dag <- grViz("
digraph mothur_snakemake_dag {
    graph[bgcolor=white, margin=0];
    node[shape=box, style=rounded, fontname=sans, fontsize=10, penwidth=2];
    edge[penwidth=2, color=grey];
    
        0[label = 'Mothur Bioinformatics Workflow', color = '0.13 0.6 0.85', style='oval'];
        1[label = 'get_zymo_mock', color = '0.50 0.6 0.85', style='rounded'];
        2[label = 'get_references', color = '0.30 0.6 0.85', style='rounded'];
        3[label = 'final_otu_files', color = '0.57 0.6 0.85', style='rounded'];
        4[label = 'classify_otu97', color = '0.23 0.6 0.85', style='rounded'];
        5[label = 'cluster_otu97', color = '0.37 0.6 0.85', style='rounded'];
        6[label = 'compute_seqdist003', color = '0.27 0.6 0.85', style='rounded'];
        7[label = 'denoise_n_classify80', color = '0.07 0.6 0.85', style='rounded'];
        8[label = 'align_n_filter', color = '0.43 0.6 0.85', style='rounded'];
        9[label = 'make_contigs', color = '0.53 0.6 0.85', style='rounded'];
        10[label = 'make_files', color = '0.17 0.6 0.85', style='rounded'];
        11[label = 'final_phylotype_files', color = '0.03 0.6 0.85', style='rounded'];
        12[label = 'classify_phylotype', color = '0.60 0.6 0.85', style='rounded'];
        13[label = 'final_asv_files', color = '0.33 0.6 0.85', style='rounded'];
        14[label = 'classify_asv', color = '0.63 0.6 0.85', style='rounded'];
        15[label = 'final_phylogeny_files', color = '0.40 0.6 0.85', style='rounded'];
        16[label = 'classify_phylogeny', color = '0.10 0.6 0.85', style='rounded'];
        17[label = 'otutable_by_group', color = '0.47 0.6 0.85', style='rounded'];
        18[label = 'subsample_otutable', color = '0.20 0.6 0.85', style='rounded'];
        19[label = 'count_groups', color = '0.00 0.6 0.85', style='rounded'];
        15 -> 0
        17 -> 0
        1 -> 0
        11 -> 0
        3 -> 0
        18 -> 0
        13 -> 0
        2 -> 1
        4 -> 3
        5 -> 4
        7 -> 4
        6 -> 5
        7 -> 5
        7 -> 6
        8 -> 7
        2 -> 7
        9 -> 8
        2 -> 8
        10 -> 9
        12 -> 11
        7 -> 12
        14 -> 13
        7 -> 14
        16 -> 15
        7 -> 16
        3 -> 17
        17 -> 18
        19 -> 18
        17 -> 19
        }")

mothur_snakemake_dag
```

<br><br>

## Microbial profiling classification options
### 1. Operational Taxonomic Units (OTUs)
- OTUs are clusters of similar sequences and are commonly accepted as analytical units in microbial profiling when using 16S rRNA gene markers.

### 2. Phylotype
- A phylotype in microbiome research is a DNA sequence or group of sequences sharing more than an arbitrarily chosen level of similarity of a 16S rRNA gene marker.

### 3. Amplicon Sequence Variant (ASV)
- An ASV in microbiome research is any inferred single DNA sequences recovered from a bioinformatics analysis of 16S rRNA marker genes.

### 4. Microbial Phylogenies
- Microbial phylogenies are from gene sequence homologies.  Models of mutation determine the most-likely evolutionary histories.

<br><br><hr width=50%>

# Appendix {-}
## Reference Databases
<ol>
  <li>Mothur-based SILVA reference files [@MothurSILVA].</li>
  <li>Mothur-based RDP reference files [@MothurRDP]. Note: The RDP database is to classify 16S rRNA gene sequences to the genus level.</li>
  <li>ZymoBIOMICS Microbial Community Standard (Cat # D6306)[@Zymo2023]. The ZymoBIOMICS Microbial Community DNA Standard is designed to assess bias, errors and other artifacts after the step of nucleic acid purification.</li>
</ol>

<br>
<br>

## Sample location (demo)

<img src="images/samplelocmap.png"/>

<br>
<br>

## Troubleshooting (in progress)
<ol>
  <li>Mothur dist.seqs taking too long.</li>
  <ul>
    <li>Merged reads are too long, probably over 300pb.</li>
    <li>Reads not overlaping when merging the paired reads.</li>
    <li>Too many uniques representative sequences probably caused by lack of overlapping.</li>
    <li>No enough computer power which suggest a use of HPC or Cluster.</li>
  </ul>
</ol>

<br>
<br>

## References
::: {#refs}
:::

<br><br><hr width=50%><br>

<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`. <br><br>
These workflows are continuously reviewed, integrated, tested and compiled by [`r rmarkdown::metadata$author`](mailto:`r rmarkdown::metadata$email_address`).  
Github_Repo: `r rmarkdown::metadata$github_repo`.  
Github_Page (this document): `r rmarkdown::metadata$github_page`.  
Visit the [Complex Data Insights (CDI) website](`r rmarkdown::metadata$related_website`) for more practical user guides (...in progress).  
</div><br><br>
