---
title:  "<big><u>IMAP-PART2</u></big><br><br>Bioinformatics Analysis of Microbiome Data<br><br>"
subtitle: |
  | ![](images/bioinformatics.png)<br>
  | Integrated Microbiome Analysis Pipelines<br>( IMAP-PART2 ) <br><br><small>[Latest GitHub-Repo](https://github.com/tmbuza/imap-bioinformatics/)</small>
date: |
  |  Updated on `r Sys.Date()`
author: Teresia Mrema-Buza
output: 
  html_document: 
    css: css/styles.css
csl: "library/apa.csl"
bibliography:
- library/references.bib
- library/imap.bib
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-bioinformatics"
related_website: "https://complexdatainsights.com"
description: |
  This repo provides integrated and highly curated solutions for reproducible research.
---
> We continuously review and integrate the old iMAP [@TMBuza2019] with Snakemake and GitHub actions to facilitate reproducible microbiome data analysis!

<br><hr width=100%><br>


<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Montserrat&display=swap" rel="stylesheet">
<link rel="preconnect" href="https://fonts.googleapis.com/css2?family=Anton&display=swap" rel="stylesheet">

```{r echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  cache = FALSE,
  comment = NA,
  fig.path='./figures/',
  fig.show='asis',
  dev = 'png',
  fig.align='center',
  out.width = "70%",
  fig.width = 7,
  fig.asp = 0.7,
  fig.show = "asis"
)

library(tidyverse, suppressPackageStartupMessages())
library(schtools)

```
<br><br>

# General Overview

- We use the snakemake workflow management system [@Koster2021; @Snakemake2023] for:
  - Maintaining reproducibility in technical validation and regeneration of results
  - Creating scalable data analysis scaled to server, grid or cloud environment seamlessly.
  - Fostering sustainable improvement of the microbiome data analysis.
- Reviewing existing workflows methodlogies [@Snakemake2023; @Mothursnakemake] helps in gaining a better insights for improving microbiome data analysis.
- We break any complex workflows into small contiguous but related chunks where each major step form a separate executable snakemake rule.

<br>

> We envision to keep fostering on continuous integration and development of highly reproducible workflows.


## Snakemake rulegraph
<img src="dags/rulegraph.svg">

<br><br>

## Project directories
```{bash}
cat project_tree.txt
```

<br><hr width=100%><br>


# Preprocessing 16S Reads
- At this point we are aware of the preprocessing and bioinformatics tools needed to getting started with microbiome data analysis. 
- We assume that the tools are already installed. Now it is time to use them to:
  - Get simple statistics.
  - Check the quality of the reads.
  - Create summary report of quality metrics.
  - Trim poor read at a user-specified cutoff.
  - Remove contaminants.

<br><br>

# Microbial Profiling Using 16S 

## Read classification options (mothur)
### Classify OTUs
- OTUs (Operational Taxonomic Units (OTUs)) are clusters of similar sequences and are commonly accepted as analytical units in microbial profiling when using 16S rRNA gene markers.

### Classify Phylotypes
- A phylotype in microbiome research is a DNA sequence or group of sequences sharing more than an arbitrarily chosen level of similarity of a 16S rRNA gene marker.

### Classify ASVs
- ASVs Amplicon Sequence Variants (ASVs)in microbiome research is any inferred single DNA sequences recovered from a bioinformatics analysis of 16S rRNA marker genes.
- ASV is typically really a cluster of sequences that are one or two bases apart from each other.

### Classify Phylogenies
- Microbial phylogenies are from gene sequence homologies.  Models of mutation determine the most-likely evolutionary histories.

<br><br>

## Mothur preliminary analysis
The preliminary analysis (`alpha_beta_diversity rule`) is part of the bioinformatics analysis. It includes:

- Creating reads count for each group.
- Subsampling for downstream analysis.
- Rarefaction.
- Computing Alpha diversity metrics.
- Computing Beta diversity metrics.
- Getting sample distances.
- Constructing sample phylip tree.
- Generating ordination matrices including PCoA and NMDS.


<br><hr width=100%><br>

# Appendix {-}
## Reference Databases
<ol>
  <li>Mothur-based SILVA reference files [@MothurSILVA]</li>.
  <li>Mothur-based RDP reference files [@MothurRDP]. Note: The RDP database is to classify 16S rRNA gene sequences to the genus level.</li>
  <li>ZymoBIOMICS Microbial Community Standard (Cat # D6306)[@Zymo2023]. The ZymoBIOMICS Microbial Community DNA Standard is designed to assess bias, errors and other artifacts after the step of nucleic acid purification.</li>
</ol>

<br>
<br>




```{r eval=FALSE, include=FALSE}
## Machine Learning
library(DiagrammeR)
library(DiagrammeRsvg)
library(magrittr)
library(rsvg)
graph <- DiagrammeR::grViz("
digraph {
graph [layout = dot, rankdir = TD]
# define the global styles of the nodes. We can override these in box if we wish
node [shape = rectangle, style = filled, fillcolor = white]
A [label = 'Clean Data', shape = folder, fillcolor = grey]
B [label = 'Training \n Set', shape = folder, fillcolor = grey]
C [label = 'Testing \n Set', shape = folder, fillcolor = grey]
D [label = 'Model \nTraining Subset\n(X-Train)']
E [label = 'Validation Subset\n(y_Train)']
F [label = 'Better Feature \nSelection']
G [label = 'Retrain \nModel']
H [label = 'Best Feature \nSelection']
I [label = 'Validated \nFeatures']
J [label = 'Best Performing \nModel', shape = rectangle]
K [label = 'Predict \nTest Set Labels', shape = rectangle]
L [label = 'Compare Known \nwith Predicted', shape = rectangle]
M [label = 'Cross \nValidation', shape = rectangle]
N [label = 'Model \nImprovement', shape = rectangle]
# edge definitions with the node IDs
{A}  -> B
{A}  -> C
{B}  -> D
{B}  -> E
{D}  -> F
{F}  -> G
{G}  -> D
{G}  -> H
{H}  -> E
{E}  -> I
{I}  -> J
{J}  -> C
{C}  -> K
{K}  -> L
{L}  -> M
{M}  -> N
}", height = 1000, width = 600)
### Visualize the graph object
graph
```
## Troubleshooting (in progress)
<ol>
  <li>Are chimeras removed by default in newer versions?</li>
    <ul>Yes. Chimeras are removed by default. You can still run the remove.seqs command without error, but it is not necessary. Remove chimera sequence explained [here](https://forum.mothur.org/t/chimera-vsearch-removes-groups-and-generates-empty-files/21342/4)</ul>.
  <li>Mothur dist.seqs taking too long.</li>
  <ul>
    <li>Merged reads are too long, probably over 300pb.</li>
    <li>Reads not overlaping when merging the paired reads.</li>
    <li>Too many uniques representative sequences probably caused by lack of overlapping.</li>
    <li>No enough computer power which suggest a use of HPC or Cluster.</li>
  </ul>
</ol>

<br><br><hr width=50%><br>

## References
::: {#refs}
:::

<br><br><hr width=100%><br>

<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`. <br><br>
The snakemake workflow mentioned in this report is continuously being reviewed, integrated, tested and compiled by [`r rmarkdown::metadata$author`](mailto:`r rmarkdown::metadata$email_address`).  
Github_Repo: `r rmarkdown::metadata$github_repo`.  

</div><br><br>
