---
title: "Integrated Microbiome<br>Analysis Pipelines<br><br>"
subtitle: "Fostering continuous improvement and integration<br><br>"
author: TMBuza
date: |
  | ![](images/new_horizon.png)<br><br>
  | Updated on `r Sys.time()` <br><br>
csl: "library/apa-numeric-superscript.csl"
bibliography: "library/references.bib"
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-mothur"
github_page: "https://tmbuza.github.io/imap-mothur/"
related_website: "https://complexdatainsights.com"
description: |
  This guide provides integrated and highly curated solutions for achieving better results.
---


<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Montserrat&display=swap" rel="stylesheet">


```{r echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  cache = FALSE,
  comment = NA,
  fig.path='./figures/',
  fig.show='asis',
  dev = 'png',
  fig.align='center',
  out.width = "70%",
  fig.width = 7,
  fig.asp = 0.7,
  fig.show = "asis"
)

library(tidyverse, suppressPackageStartupMessages())
library(mikropml)
library(schtools)

```



```{css echo = FALSE}
/* ############
# CONTAINER CSS
############ */

.author, .subtitle{
  display: none;
} 
.main-container {
  max-width: 80%;
  background-color: black;
  color: white;
  padding: 3% 3% 3% 3%;
  margin-left: auto;
  margin-right: auto;
}
body{
  font-size: 14px;
  text-align: left;
}

#header {
  background-color: #013011;
  color: white;
  font-family: "Montserrat", sans-serif;
  text-align: center;
  width: 100%;
  padding: 5% 0% 0% 0%;
  text-transform: none;
  font-weight: bold;
  background-image: url("images/bkgd.png");
  background-repeat: no-repeat, repeat;
  background-color: none;
  height: 100%; /* You must set a specified height */
  background-position: center; /* Center the image */
  background-size: cover; /* Resize the background image to cover the entire container */
  
}
```

```{css echo = FALSE} 
/* ############
# BODY CSS
############ */
#header h1{
  font-size: 4em;
  text-transform: capitalize;
  color: black;
}
#header h2{
  font-size: 3em;
}

#header h3{
  font-size: 2.5em;
  color: black;
} 
#header img{
width:80%;
}
body h1{
  font-size: 2.5em;
  text-transform: capitalize;
}
body h2{
  font-size: 1.8em;
}

body h3{
  font-size: 1.6em;
}

body {
  background-color: F5F5F5;
  color: white;
  font-family: "Montserrat", sans-serif;
}
body h1{
  font-size: 2.5em;
  text-transform: capitalize;
}
body h2{
  font-size: 2.0em;
}

body h3{
  font-size: 1.6em;
}

table tbody tr td {
  font-size: 14px;
  text-transform: capitalize;
}
body img{
  width: 100%;
  display: block;
  margin-left: auto;
  margin-right: auto;
  padding: 20px, 50px, 20px, 50px;
}
#glimpse{
  text-align: center;
}

#glimpse h1{
  text-transform: none;
}

a{
  color: #6bb7f5;
}
#tablefield {
  max-width: 80%;
  margin-left: auto;
  margin-right: auto;
}
hr {
  display: inline-block;
  text-align: left;
  margin: 0px;
}
#glimpse hr {
  display: inline-block;
  margin: 0px;
}

```


```{css echo = FALSE} 
/* ############
# MEDIA: max-width: 768px
############ */
@media only screen and (max-width: 768px) {
  .main-container {
  max-width: 100%;
  padding: 1% 1% 1% 1%;
  margin-left: auto;
  margin-right: auto;
  }

  #header img{
  max-width: 80%;
  }
  #header h1{
    font-size: 2.5em;
    text-transform: capitalize;
    color: black;
  }

   body {
    font-size: 12px;
    width: 100%; 
  } 
  #footer {
    text-align: left;
    font-size: 1.1em;
  } 

  #tablefield {
  width: 100%;
  margin-left: auto;
  margin-right: auto;
} 
body h1{
  font-size: 2.0em;
  text-transform: capitalize;
}
body h2{
  font-size: 1.6em;
}

body h3{
  font-size: 1.6em;
}

#tablefield h2{
  font-size: 1.6em;
}

}
```
<br><br>

# Microbial Profiling

## Quick Glimpse
- We use snakemake workflow engine [@Koster2021] to maintain reproducibility which is important in technical validation, regeneration of results and improvement of the microbiome bioinformatics analysis. 

## Reviewed and improved snakemake workflows
<ol>
  <li>Mothur 16S v4 Analysis Pipeline [@mothursnakemake].</li>
  <li></li>
  <li></li>
</ol>


## Classification options
### Operational Taxonomic Units (OTUs)
- OTUs are clusters with sequence similarity and are accepted as an analytical unit used in microbial profiling using 16S rRNA gene markers.

### Phylotype
- A phylotype in microbiome research is a DNA sequence or group of sequences sharing more than an arbitrarily chosen level of similarity of a 16S rRNA gene marker.

### Amplicon Sequence Variant (ASV)
- An ASV in microbiome research is any one of the inferred single DNA sequences recovered from a bioinformatics analysis of 16S rRNA marker genes.

### Microbial Phylogenies
- Microbial phylogenies are estimated from gene sequence homologies and the most-likely evolutionary histories are commonly inferred using models of mutation.


<br><br>
<hr width="50%">
<br><br> 

# Improved Mothur Snakemake Workflow

<img src="dags/rulegraph.svg"/>

<br>
<br>

# Example of common FAQs (in progress)
<ol>
  <li>Mothur dist.seqs taking too long.</li>
  <ul>
    <li>Length of the merged reads (merged reads over 300pb).</li>
    <li>Not enough overlaping when merging the paired reads.</li>
    <li>Use other pre-cluster algorithm like Deblur.</li>
    <li>Too many uniques?</li>
    <li>Not enough computer power?: Use HPC or Cluster.</li>
    <li>Try to make contigs using maxee instead of deltaq or use phylogeny approach for downstream analysis, or use clister.split based on taxonomy.</li>
    <li>do not forget to use controls.</li>
  </ul>
</ol>


<br>
<hr width=50%>

<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`. <br><br>
These workflows are continuously reviewed, integrated, tested and compiled by [Teresia Mrema-Buza](mailto:`r rmarkdown::metadata$email_address`).  
Github_Repo: `r rmarkdown::metadata$github_repo`.  
Github_Page (this document): `r rmarkdown::metadata$github_page`.  
Visit the [Complex Data Insights (CDI) website](`r rmarkdown::metadata$related_website`) for more practical user guides (...in progress).  
</div>

<br>
<br>

## References
