---
title:  "<big><u>IMAP-PART 05</u></big><br><br>Microbial Profiling using 16S Ribosomal RNA Gene Amplicons<br><br>"
subtitle: |
  | ![](images/metadata.png)<br>
  | Fostering Reproducible Microbiome data Analysis with Snakemake workflow
date:
- IMAP-PART 01<br>[Latest GitHub-Repo](https://github.com/tmbuza/imap-bioinformatics-analysis/)<br>
- Updated on `r Sys.Date()`
author: Teresia Mrema-Buza
output: 
  html_document: 
css:
- styles.css
csl: "library/apa.csl"
bibliography:
- library/references.bib
- library/imap.bib
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-bioinformatics-analysis"
description: |
  Minimal summary describing features for getting started with microbiome data analysis.
---

<div id="main-container">

<!-- Google Fonts: https://fonts.google.com/?preview.text=IMAP%20template&preview.text_type=custom-->
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Anton" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@100;300;400;500;700,900&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Oswald:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Merriweather:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@100;200;300;400;700&display=swap" rel="stylesheet">

<span class="fas fa-my-custom-icon"></span>

```{r echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo  =FALSE,
  message  =FALSE,
  warning  =FALSE,
  cache  =FALSE,
  comment  =NA,
  fig.path='./figures/',
  fig.show='asis',
  dev  ='png',
  fig.align='center',
  out.width  ="70%",
  fig.width  =7,
  fig.asp  =0.7,
  fig.show  ="asis"
)

library(tidyverse, suppressPackageStartupMessages())
library(schtools)
```

# Microbial Profiling Using Mothur

##  Current snakemake workflow
- Typically the snakemake workflow is defined by specifying rules.
- The rule-graph graphically shows the input-output files. 
- The snakemake is capable of automatically determining the dependencies between the rules and creates a dot-like DAG (Directed Acyclic Graph).

```{bash ruledag, include=FALSE}
bash workflow/scripts/rules_dag.sh
```

<img src="dags/rulegraph.svg">

<br><hr width=100%><br>


## Mothur reference databases
<ol>
  <li>Mothur-based SILVA reference files [@MothurSILVA].</li>
  <li>Mothur-based RDP reference files [@MothurRDP]. Note: The RDP database is to classify 16S rRNA gene sequences to the genus level.</li>
  <li>ZymoBIOMICS Microbial Community Standard (Cat # D6306)[@Zymo2023]. The ZymoBIOMICS Microbial Community DNA Standard is designed to assess bias, errors and other artifacts after the step of nucleic acid purification.</li>
</ol>

## Mothur classification methods
There are four methods that can be used to profile microbial communities present in a sample. Here we briefly decribe each method:

### 1.Classify OTUs
- OTUs (Operational Taxonomic Units (OTUs)) are clusters of similar sequences and are commonly accepted as analytical units in microbial profiling when using 16S rRNA gene markers.

### 2. Classify Phylotypes
- A phylotype in microbiome research is a DNA sequence or group of sequences sharing more than an arbitrarily chosen level of similarity of a 16S rRNA gene marker.

### 3. Classify ASVs
- ASVs Amplicon Sequence Variants (ASVs)in microbiome research is any inferred single DNA sequences recovered from a bioinformatics analysis of 16S rRNA marker genes.
- ASV is typically really a cluster of sequences that are one or two bases apart from each other.

### 4. Classify Phylogenies
- Microbial phylogenies are from gene sequence homologies.  Models of mutation determine the most-likely evolutionary histories.

<br><br>

## Preliminary analysis using `Mothur`
The preliminary analysis (`alpha_beta_diversity rule`) is part of the bioinformatics analysis. It includes:

- Creating reads count for each group.
- Subsampling for downstream analysis.
- Rarefaction.
- Computing Alpha diversity metrics.
- Computing Beta diversity metrics.
- Getting sample distances.
- Constructing sample phylip tree.
- Generating ordination matrices including PCoA and NMDS.


<br><hr width=100%><br>


# Related work
```{block imaprepos, echo=TRUE}

| Repo | Description | Summary |
|--------------|----------------------------------|--------------|
| [IMAP-PART1](https://github.com/tmbuza/imap-basic-requirements/) | Getting started with microbiome data analysis | [GH-Pages](https://tmbuza.github.io/imap-basic-requirements/) |
| [IMAP-PART2](https://github.com/tmbuza/imap-sample-metadata/) | Exploring and profiling microbiome sample metadata | [GH-Pages](https://tmbuza.github.io/imap-sample-metadata/) |
| [IMAP-PART3](https://github.com/tmbuza/imap-bioinformatics-analysis/) | Bioinformatics analysis of microbiome data | [GH-Pages](https://tmbuza.github.io/imap-bioinformatics-analysis/) |
| [IMAP-PART4](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART5](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART6](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART7](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART8](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
```


## Citation

> Please consider citing the iMAP article [@TMBuza2019] if you find any part of the IMAP practical user guides helpful in your microbiome data analysis.

# References
::: {#refs}
:::

<br><hr width=100%><br>

# Appendix {-}

## Project main tree
```{bash, include=FALSE}
bash workflow/scripts/tree.sh
```

```{bash treeimg}
cat results/project_tree.txt
```

## Screenshot of interactive snakemake report {#smkreport}
> The interactive snakemake html report can be viewed by opening the `report.html` using any compartible browser. You will be able to explore the workflow and the associated statistics. You will also be able to close the left bar to get a better wider view of the display.

```{bash smkrpt, include=FALSE}
snakemake --unlock;
snakemake --report report.html;
hti -H report.html -o images/smkreport;
```

```{bash staticsmk, include=FALSE}
snakemake --report report.html
bash workflow/scripts/smk_html_report.sh
```

<img src="images/smkreport/screenshot.png">


## Troubleshooting of FAQs
<ol>
  <li>Are chimeras removed by default in the newer versions of mothur?</li>
    <ul>Yes! Chimeras are removed by default. You can still run the remove.seqs command without error, but it is not necessary. Remove chimera sequence explained [here](https://forum.mothur.org/t/chimera-vsearch-removes-groups-and-generates-empty-files/21342/4)</ul>.
  <li>Wny Mothur dist.seqs is taking too long?</li>
  <ul>
    <li>Merged reads are too long, probably over 300pb.</li>
    <li>Reads not overlaping when merging the paired reads.</li>
    <li>Too many uniques representative sequences probably caused by lack of overlapping.</li>
    <li>No enough computer power which suggest a use of HPC or Cluster.</li>
  </ul>
  <li>Question</li>
    <ul>Answer</ul>
</ol>


</div> <!-- end main-container -->

<br><hr width=100%><br>

<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`.<br>
Github_Repo: `r rmarkdown::metadata$github_repo`.<br>
Repo compiled and maintained by: [`r rmarkdown::metadata$author`](mailto:`r rmarkdown::metadata$email_address`). 
</div>

